File: src/turtlebot/navigation/map_navigation/scripts/map_navigation.py

import rospy
import actionlib
from move_base_msgs.msg import MoveBaseAction, MoveBaseGoal
from math import radians, degrees
from actionlib_msgs.msg import *
from geometry_msgs.msg import Point 
from sound_play.libsoundplay import SoundClient

class map_navigation():
	
	def choose(self):

		choice='q'
		
		rospy.loginfo("|-------------------------------|")
		rospy.loginfo("|PRESSE A KEY:")
		rospy.loginfo("|'0': Cafe ")
		rospy.loginfo("|'1': Office 1 ")
		rospy.loginfo("|'2': Office 2 ")
		rospy.loginfo("|'3': Office 3 ")
		rospy.loginfo("|'q': Quit ")
		rospy.loginfo("|-------------------------------|")
		rospy.loginfo("|WHERE TO GO?")
		choice = input()
		return choice

	def __init__(self): 

		sc = SoundClient()
		path_to_sounds = "/home/ros/catkin_ws/src/gaitech_edu/src/sounds/"

		# declare the coordinates of interest 
		self.xCafe =  7.31
		self.yCafe = 1.46
		self.xOffice1 = 5.10
		self.yOffice1 = -3.82
		self.xOffice2 = -2.35
		self.yOffice2 = -0.95
		self.xOffice3 = 35.20
		self.yOffice3 = 13.50
		self.goalReached = False
		# initiliaze
        	rospy.init_node('map_navigation', anonymous=False)
		choice = self.choose()
		
		if (choice == 0):

			self.goalReached = self.moveToGoal(self.xCafe, self.yCafe)
		
		elif (choice == 1):

			self.goalReached = self.moveToGoal(self.xOffice1, self.yOffice1)

		elif (choice == 2):
			
			self.goalReached = self.moveToGoal(self.xOffice2, self.yOffice2)
		
		elif (choice == 3):

			self.goalReached = self.moveToGoal(self.xOffice3, self.yOffice3)

		if (choice!='q'):

			if (self.goalReached):
				rospy.loginfo("Congratulations!")
				#rospy.spin()

				sc.playWave(path_to_sounds+"ship_bell.wav")
				
				#rospy.spin()

			else:
				rospy.loginfo("Hard Luck!")
				sc.playWave(path_to_sounds+"short_buzzer.wav")
		
		while choice != 'q':
			choice = self.choose()
			if (choice == 0):

				self.goalReached = self.moveToGoal(self.xCafe, self.yCafe)
		
			elif (choice == 1):

				self.goalReached = self.moveToGoal(self.xOffice1, self.yOffice1)

			elif (choice == 2):
		
				self.goalReached = self.moveToGoal(self.xOffice2, self.yOffice2)
		
			elif (choice == 3):

				self.goalReached = self.moveToGoal(self.xOffice3, self.yOffice3)

			if (choice!='q'):

				if (self.goalReached):
					rospy.loginfo("Congratulations!")
					#rospy.spin()

					sc.playWave(path_to_sounds+"ship_bell.wav")

				else:
					rospy.loginfo("Hard Luck!")
					sc.playWave(path_to_sounds+"short_buzzer.wav")


	def shutdown(self):
        # stop turtlebot
        	rospy.loginfo("Quit program")
        	rospy.sleep()

	def moveToGoal(self,xGoal,yGoal):

		#define a client for to send goal requests to the move_base server through a SimpleActionClient
		ac = actionlib.SimpleActionClient("move_base", MoveBaseAction)

		#wait for the action server to come up
		while(not ac.wait_for_server(rospy.Duration.from_sec(5.0))):
			rospy.loginfo("Waiting for the move_base action server to come up")
		

		goal = MoveBaseGoal()

		#set up the frame parameters
		goal.target_pose.header.frame_id = "map"
		goal.target_pose.header.stamp = rospy.Time.now()

		# moving towards the goal*/

		goal.target_pose.pose.position =  Point(xGoal,yGoal,0)
		goal.target_pose.pose.orientation.x = 0.0
		goal.target_pose.pose.orientation.y = 0.0
		goal.target_pose.pose.orientation.z = 0.0
		goal.target_pose.pose.orientation.w = 1.0

		rospy.loginfo("Sending goal location ...")
		ac.send_goal(goal)

		ac.wait_for_result(rospy.Duration(60))

		if(ac.get_state() ==  GoalStatus.SUCCEEDED):
			rospy.loginfo("You have reached the destination")	
			return True
	
		else:
			rospy.loginfo("The robot failed to reach the destination")
			return False

if __name__ == '__main__':
    try:
	
	rospy.loginfo("You have reached the destination")
        map_navigation()
        rospy.spin()
    except rospy.ROSInterruptException:
        rospy.loginfo("map_navigation node terminated.")

File: src/RIA-E100/navigation/map_navigation.cpp
/*
 *  Gaitech Educational Portal
 *
 * Copyright (c) 2016
 * All rights reserved.
 *
 * License Type: GNU GPL
 *
 *   This program is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation, either version 3 of the License, or
 *   (at your option) any later version.
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 *   Program: Map-Based Navigation
 *
 */

#include <ros/ros.h>
#include <move_base_msgs/MoveBaseAction.h>
#include <actionlib/client/simple_action_client.h>
#include "sound_play/sound_play.h"

std::string path_to_sounds;

/** function declarations **/
bool moveToGoal(double xGoal, double yGoal);
char choose();

/** declare the coordinates of interest **/
double xCafe = 15.50;
double yCafe = 10.20;
double xOffice1 = 27.70 ;
double yOffice1 = 13.50;
double xOffice2 = 30.44 ;
double yOffice2 = 13.50;
double xOffice3 = 35.20 ;
double yOffice3 = 13.50;

bool goalReached = false;

int main(int argc, char** argv){
	ros::init(argc, argv, "map_navigation_node");
	ros::NodeHandle n;
	sound_play::SoundClient sc;
	ros::spinOnce();
	path_to_sounds = "/home/ros/catkin_ws/src/gaitech_edu/src/sounds/";
	//sc.playWave(path_to_sounds+"short_buzzer.wav");
	//tell the action client that we want to spin a thread by default

	char choice = 'q';
	do{
		choice =choose();
		if (choice == '0'){
			goalReached = moveToGoal(xCafe, yCafe);
		}else if (choice == '1'){
			goalReached = moveToGoal(xOffice1, yOffice1);
		}else if (choice == '2'){
			goalReached = moveToGoal(xOffice2, yOffice2);
		}else if (choice == '3'){
			goalReached = moveToGoal(xOffice3, yOffice3);
		}
		if (choice!='q'){
			if (goalReached){
				ROS_INFO("Congratulations!");
				ros::spinOnce();
				sc.playWave(path_to_sounds+"ship_bell.wav");
				ros::spinOnce();

			}else{
				ROS_INFO("Hard Luck!");
				sc.playWave(path_to_sounds+"short_buzzer.wav");
			}
		}
	}while(choice !='q');
	return 0;
}

bool moveToGoal(double xGoal, double yGoal){

	//define a client for to send goal requests to the move_base server through a SimpleActionClient
	actionlib::SimpleActionClient<move_base_msgs::MoveBaseAction> ac("move_base", true);

	//wait for the action server to come up
	while(!ac.waitForServer(ros::Duration(5.0))){
		ROS_INFO("Waiting for the move_base action server to come up");
	}

	move_base_msgs::MoveBaseGoal goal;

	//set up the frame parameters
	goal.target_pose.header.frame_id = "map";
	goal.target_pose.header.stamp = ros::Time::now();

	/* moving towards the goal*/

	goal.target_pose.pose.position.x =  xGoal;
	goal.target_pose.pose.position.y =  yGoal;
	goal.target_pose.pose.position.z =  0.0;
	goal.target_pose.pose.orientation.x = 0.0;
	goal.target_pose.pose.orientation.y = 0.0;
	goal.target_pose.pose.orientation.z = 0.0;
	goal.target_pose.pose.orientation.w = 1.0;

	ROS_INFO("Sending goal location ...");
	ac.sendGoal(goal);

	ac.waitForResult();

	if(ac.getState() == actionlib::SimpleClientGoalState::SUCCEEDED){
		ROS_INFO("You have reached the destination");
		return true;
	}
	else{
		ROS_INFO("The robot failed to reach the destination");
		return false;
	}

}

char choose(){
	char choice='q';
	std::cout<<"|-------------------------------|"<<std::endl;
	std::cout<<"|PRESSE A KEY:"<<std::endl;
	std::cout<<"|'0': Cafe "<<std::endl;
	std::cout<<"|'1': Office 1 "<<std::endl;
	std::cout<<"|'2': Office 2 "<<std::endl;
	std::cout<<"|'3': Office 3 "<<std::endl;
	std::cout<<"|'q': Quit "<<std::endl;
	std::cout<<"|-------------------------------|"<<std::endl;
	std::cout<<"|WHERE TO GO?";
	std::cin>>choice;

	return choice;


}
